### PR TITLE
feat(storage): Save genesis and its metadata in storage

### PR DESCRIPTION
**Pull request by msbrogli**
Migrated from [MR!385](https://gitlab.com/HathorNetwork/hathor-python/-/merge_requests/385)

### CODE REVIEW

File: hathor/transaction/storage/remote_storage.py [hunk 1]
```diff
@@ -534,13 +533,11 @@ def get_tx_count(self) -> int:
...
     def get_genesis(self, hash_bytes: bytes) -> Optional['BaseTransaction']:
-        if not self._genesis_cache:
-            self._create_genesis_cache()
+        assert self._genesis_cache is not None
         return self._genesis_cache.get(hash_bytes, None)
 
     def get_all_genesis(self) -> Set['BaseTransaction']:
-        if not self._genesis_cache:
-            self._create_genesis_cache()
+        assert self._genesis_cache is not None
```
Comment (obiyankenobi) [inline hathor/transaction/storage/remote_storage.py:540]: Isn't `self._genesis_cache` removed?

Comment (pedroferreira1) [reply hathor/transaction/storage/remote_storage.py:540]: The genesis are being added to `_genesis_cache ` on `_save_or_verify_genesis`.

File: tests/tx/test_genesis.py [hunk 2]
```diff
@@ -20,39 +20,42 @@ def get_genesis_output():
...
 
 class GenesisTest(unittest.TestCase):
-    def test_pow(self):
-        genesis = get_genesis_transactions(None)
+    def setUp(self):
+        super().setUp()
+        self.storage = TransactionMemoryStorage()
 
+    def test_pow(self):
+        genesis = self.storage.get_all_genesis()
```
Comment (pedroferreira1) [inline tests/tx/test_genesis.py:28]: You could have this line on the `setUp` method, so you don't repeat it in all the tests.

File: tests/tx/test_tx.py [hunk 3]
```diff
@@ -517,7 +517,6 @@ def test_tx_methods(self):
 
         # Validate is_genesis without storage
         tx_equal.storage = None
-        self.assertFalse(tx_equal.is_genesis)
```
Comment (pedroferreira1) [inline tests/tx/test_tx.py:520]: Why did you remove it? Maybe we should just check if this raises `ValueError`

File: hathor/transaction/block.py [hunk 4]
```diff
@@ -123,7 +123,7 @@ def create_from_struct(cls, struct_bytes: bytes,
 
     def calculate_height(self) -> int:
         """Return the height of the block, i.e., the number of blocks since genesis"""
-        if self.is_genesis:
+        if not self.parents:
```
[the remark this answers is not in the document]
Comment (msbrogli) [reply hathor/transaction/block.py:126]: The `is_genesis` is a method, so I had to use another verification to skip the genesis. This is one of the dangers of using `@property` decorator.

File: hathor/transaction/genesis.py [hunk 5]
```diff
@@ -62,7 +62,9 @@ def _get_genesis_hash() -> bytes:
 GENESIS_HASH = _get_genesis_hash()
 
 
-def get_genesis_transactions(tx_storage: Optional[TransactionStorage]) -> List[BaseTransaction]:
+def get_genesis_transactions(tx_storage: Optional[TransactionStorage], safe: bool = False) -> List[BaseTransaction]:
+    if not safe:
+        raise RuntimeError('You shouldn\'t get genesis directly. Please, get it from your storage instead.')
```
[the remark this answers is not in the document]
Comment (msbrogli) [reply hathor/transaction/genesis.py:67]: The idea behind it is that you should not use it unless you know what you're doing. I don't think the `get_genesis_transactions_safe` is clear that you probably shouldn't use it.

File: tests/tx/test_tx.py (hunk 3 above)
Comment (msbrogli) [reply tests/tx/test_tx.py:520]: The `is_genesis` is a method marked as `@property`. It does not work anymore without a storage.

File: tests/tx/test_genesis.py (hunk 2 above)
Comment (msbrogli) [reply tests/tx/test_genesis.py:28]: They are independent test but it happened that all of them need the genesis. We have plenty of things in the `setUp` of other tests because all tests used those things but there are some new tests that doesn't. But I can move there if we agree that's the best place to be.

File: hathor/cli/main.py [hunk 6]
```diff
@@ -138,8 +138,8 @@ def main():
     except KeyboardInterrupt:
         logger.warn('Aborting and exiting...')
         sys.exit(1)
-    except Exception:
-        logger.exception('Uncaught exception:')
+    except Exception as e:
+        logger.exception('Uncaught exception:', e)
```
[the remark this answers is not in the document]
Comment (msbrogli) [reply hathor/cli/main.py:142]: I agree. I just rolled back my change. Thanks!

File: hathor/transaction/genesis.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (obiyankenobi) [reply hathor/transaction/genesis.py:67]: In that sense, I don't think a parameter `safe=True` helps. Maybe starting the method with `_` to indicate it's "internal".

[the remark this answers is not in the document]
Comment (msbrogli) [reply hathor/transaction/genesis.py:67]: I agree that `safe` may not be the best parameter. Maybe `i_know_what_im_doing=True` parameter?

[the remark this answers is not in the document]
Comment (obiyankenobi) [reply hathor/transaction/genesis.py:67]: Still seems weird to me. We have several methods that should be used with extreme caution - e.g.`_add_to_cache()`. In none of them we have such a parameter.

File: tests/tx/test_genesis.py (hunk 2 above)
Comment (pedroferreira1) [reply tests/tx/test_genesis.py:28]: I am okay keeping it the way it is. The idea was just to prevent code duplication.

File: tests/tx/test_tx.py (hunk 3 above)
Comment (pedroferreira1) [reply tests/tx/test_tx.py:520]: Yes, that's why I said we could check for a `ValueError` because that's what you raise when there is no storage.

File: hathor/transaction/storage/transaction_storage.py [hunk 7]
```diff
@@ -575,7 +600,7 @@ def _save_transaction(self, tx: BaseTransaction, *, only_metadata: bool = False)
     def _reset_cache(self) -> None:
         """Reset all caches. This function should not be called unless you know what you are doing."""
         if not self.with_index:
-            raise NotImplementedError
+            raise RuntimeError('Cannot reset cache because it has not been enabled.')
```
Comment (jansegre) [inline hathor/transaction/storage/transaction_storage.py:603]: Why not just `assert self.with_index, 'Cannot reset cache because it has not been enabled.'`? Is this a situation we can actually reach except from a miss-configured code setup?

File: hathor/transaction/genesis.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jansegre) [reply hathor/transaction/genesis.py:67]: `safe=True` actually gives me the wrong idea, I understood you could only do that if `safe` is turned on, as in there is no way to turn a manual unsafe mode that would skip something. For me a parameter does not make sense. I'd rather it be called `_get_genesis_transactions`, `_get_genesis_transactions_unsafe` or `get_genesis_transactions_unsafe` as in "it is unsafe to call this method if you don't know what you're doing"

I'm pretty sure we have many other methods that can only be called in specific situations and they're all either implicitly checked (like having to call `.start()` on managers/listeners/...) or prefixed with an underscore (like `TransactionStorage._manually_initialize`). I think adding another way to mark a method that is unsafe to call unless you know the context will just make it harder to understand what's actually going on. An underscore would be enough for me, but if we want to be extra careful and `_unsafe` suffix makes a lot of sense to me.

File: tests/tx/test_genesis.py (hunk 2 above)
Comment (jansegre) [reply tests/tx/test_genesis.py:28]: I mean, `setUp` is just to make it practical, we could have another `TestCase`, but it's just so few tests that it's maybe not worth the line count increase.

File: tests/tx/test_tx.py (hunk 3 above)
Comment (jansegre) [reply tests/tx/test_tx.py:520]: Why does it need a storage? We should be able to tell from looking at the settings only.

File: hathor/transaction/block.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jansegre) [reply hathor/transaction/block.py:126]: Being a property shouldn't matter much, but from a comment below I get that it's because it uses the storage, but I think we could do away with depending on the storage, since using the storage is being more trouble than being useful.

File: hathor/transaction/genesis.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (msbrogli) [reply hathor/transaction/genesis.py]: I agree. Can you review it again, please? I kept the changes in a different commit.

File: hathor/transaction/block.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (msbrogli) [reply hathor/transaction/block.py:126]: I didn't understand your suggestion. I agree that it would be better to use `if self.is_genesis:`, and I'll check what part breaks when I do it.

File: tests/tx/test_tx.py (hunk 3 above)
Comment (msbrogli) [reply tests/tx/test_tx.py:520]: One of the ideas of this PR is to use the storage as the point of truth regarding genesis. Maybe we can create a `is_genesis(tx_id: bytes)` method in `hathor.transaction.genesis`, and use it directly.

File: hathor/transaction/storage/transaction_storage.py (hunk 7 above)
Comment (msbrogli) [reply hathor/transaction/storage/transaction_storage.py:603]: Exactly. That's why I chose RuntimeError, but we can use an assert statement as well.

Comment (msbrogli) [reply hathor/transaction/storage/transaction_storage.py]: Changed! Can you check again, please?

File: hathor/transaction/storage/transaction_storage.py [hunk 8]
```diff
@@ -549,17 +568,23 @@ class BaseTransactionStorage(TransactionStorage):
     def __init__(self, with_index: bool = True, pubsub: Optional[Any] = None) -> None:
         super().__init__()
 
+        # Pubsub is used to publish tx voided and winner but it's optional
+        self.pubsub = pubsub
```
Comment (obiyankenobi) [inline hathor/transaction/storage/transaction_storage.py:572]: Already done on line 580.

File: hathor/transaction/block.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jansegre) [reply hathor/transaction/block.py:126]: What I suggested was:

```python
if not self.parents:
    assert self.is_genesis
    ...
```

But now I understand that the change of the `if` condition from `self.is_genesis` to `not self.parents` is precisely because `self.is_genesis` needs a storage. If that's the case, fixing `self.is_genesis` seems like it would make it possible to use that for the `if` condition. Which makes a little more sense to me, but it's probably subjective anyway. Looking back it doesn't seem that important.

File: hathor/transaction/genesis.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jansegre) [reply hathor/transaction/genesis.py]: Looks great now.

File: tests/tx/test_tx.py (hunk 3 above)
Comment (jansegre) [reply tests/tx/test_tx.py:520]: While I think it's generally useful to contain the concerns, even the storage will have to rely on the settings to know what are the genesis transactions. So I don't think it's useful to "prevent" these settings from being useful outside the storage.

I'd say that the storage can definitely be the gatekeeper of all metadata, but determining if a transaction is part of the genesis isn't inherently something that should depend on the metadata.

 And because there actually are use cases (like this assert or `calculate_height`'s `if`) that are prevented from using `is_genesis` because the lack of a storage would break it. I think it's more useful not to use the storage when it makes sense.

File: hathor/transaction/storage/transaction_storage.py (hunk 7 above)
Comment (jansegre) [reply hathor/transaction/storage/transaction_storage.py]: Looks good.

File: hathor/transaction/block.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (msbrogli) [reply hathor/transaction/block.py:126]: I agree. I just reverted the change. Would you check again, please?

File: tests/tx/test_tx.py (hunk 3 above)
Comment (msbrogli) [reply tests/tx/test_tx.py:520]: I agree. I reverted this change. Would you review it again, please?

File: hathor/transaction/storage/transaction_storage.py (hunk 8 above)
Comment (msbrogli) [reply hathor/transaction/storage/transaction_storage.py:572]: Fixed. Thanks!

File: tests/tx/test_tx.py (hunk 3 above)
Comment (pedroferreira1) [reply tests/tx/test_tx.py]: Good to me

Comment (jansegre) [reply tests/tx/test_tx.py]: Great!

File: hathor/transaction/block.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jansegre) [reply hathor/transaction/block.py]: Great!
